source/connection_manager.c

```diff
@@ -523,11 +527,59 @@ static void s_aws_connection_management_transaction_init(
     aws_linked_list_init(&work->completions);
     work->manager = manager;
     work->allocator = manager->allocator;
+    aws_ref_count_acquire(&manager->internal_ref_count);
 }
 
 static void s_aws_connection_management_transaction_clean_up(struct aws_connection_management_transaction *work) {
     AWS_FATAL_ASSERT(aws_linked_list_empty(&work->connections_to_release));
     AWS_FATAL_ASSERT(aws_linked_list_empty(&work->completions));
+    AWS_ASSERT(work->manager);
+    aws_ref_count_release(&work->manager->internal_ref_count);
+}
+
+/* The count acquire and release all needs to be invoked helding the lock */
+static void s_connection_manager_count_increase(
+    struct aws_http_connection_manager *manager,
+    enum aws_http_connection_manager_count_type count_type,
+    size_t num) {
+    switch (count_type) {
+        case AWS_HCMCT_VENDED_CONNECTION:
+            manager->vended_connection_count += num;
+            break;
+        case AWS_HCMCT_PENDING_CONNECTIONS:
+            manager->pending_connects_count += num;
+            break;
+        case AWS_HCMCT_OPEN_CONNECTION:
+            manager->open_connection_count += num;
+            break;
+        default:
+            AWS_ASSERT(false);
+    }
+    for (size_t i = 0; i < num; i++) {
+        aws_ref_count_acquire(&manager->internal_ref_count);
+    }
+}
+
+static void s_connection_manager_count_decrease(
```

Comment:
commenting that this should only ever be called during a transaction, and it can never be the thing that brings refounct to 0